bindings/rust/extended/s2n-tls/src/external_psk.rs

```diff
@@ -0,0 +1,239 @@
+        unsafe {
+            s2n_psk_set_identity(
+                self.psk.as_s2n_ptr_mut(),
+                identity.as_ptr(),
+                identity_length,
+            )
+            .into_result()
+        }?;
+        self.has_identity = true;
+        Ok(self)
+    }
+
+    /// Set the PSK secret.
+    ///
+    /// Secrets must be at least 16 bytes.
+    ///
+    /// Corresponds to [`s2n_psk_set_secret`].
+    pub fn with_secret(&mut self, secret: &[u8]) -> Result<&mut Self, crate::error::Error> {
+        let secret_length = secret.len().try_into().map_err(|_| {
+            Error::bindings(
+                ErrorType::UsageError,
+                "invalid psk secret",
+                "The secret must be no longer than u16::MAX",
+            )
+        })?;
+
+        // These checks are only in the Rust code. Adding them to C would be a
+        // backwards incompatible change.
+        //= https://www.rfc-editor.org/rfc/rfc9257.html#section-6
+        //# Each PSK ... MUST be at least 128 bits long
+        if secret_length < (128 / 8) {
+            return Err(Error::bindings(
+                ErrorType::UsageError,
+                "invalid psk secret",
+                "PSK secret must be at least 128 bits",
+            ));
+        }
+        unsafe {
+            s2n_psk_set_secret(self.psk.as_s2n_ptr_mut(), secret.as_ptr(), secret_length)
+                .into_result()
+        }?;
+        self.has_secret = true;
+        Ok(self)
+    }
+
+    /// Set the HMAC function associated with the PSK.
+    ///
+    /// Corresponds to [`s2n_psk_set_hmac`].
+    pub fn with_hmac(&mut self, hmac: PskHmac) -> Result<&mut Self, crate::error::Error> {
+        unsafe { s2n_psk_set_hmac(self.psk.as_s2n_ptr_mut(), hmac.into()).into_result() }?;
+        self.has_hmac = true;
+        Ok(self)
+    }
+
+    pub fn build(self) -> Result<ExternalPsk, crate::error::Error> {
+        if !self.has_identity {
+            Err(Error::bindings(
+                crate::error::ErrorType::UsageError,
+                "invalid psk",
+                "You must set an identity using `with_identity`",
+            ))
+        } else if !self.has_secret {
+            Err(Error::bindings(
+                crate::error::ErrorType::UsageError,
+                "invalid psk",
+                "You must set a secret using `with_secret`",
+            ))
+        } else if !self.has_hmac {
+            Err(Error::bindings(
+                crate::error::ErrorType::UsageError,
+                "invalid psk",
+                "You must set an hmac `with_hmac`",
+            ))
+        } else {
+            Ok(self.psk)
+        }
+    }
+}
+
+crate::foreign_types::define_owned_type!(
+    /// ExternalPsk represents an out-of-band pre-shared key.
+    ///
+    /// If two peers already have some mechanism to securely exchange secrets, then
+    /// they can use ExternalPSKs to authenticate rather than certificates.
+    pub ExternalPsk,
+    s2n_psk
+);
```

Comment:
Idk, I'm not sure how much benefit we really get from this macro. It doesn't seem to simplify much, but it does make what's happening less obvious.